pep-0688.rst

```diff
@@ -0,0 +1,258 @@
+---------------
+
+There are two current workarounds for annotating buffer types in
+the type system, but neither is adequate.
+
+First, the `current workaround <https://github.com/python/typeshed/blob/2a0fc1b582ef84f7a82c0beb39fa617de2539d3d/stdlib/_typeshed/__init__.pyi#L194>`__
+for buffer types in typeshed is a type alias
+that lists well-known buffer types in the standard library, such as
+``bytes``, ``bytearray``, ``memoryview``, and ``array.array``. This
+approach works for the standard library, but it does not extend to
+third-party buffer types.
+
+Second, the `documentation <https://docs.python.org/3.10/library/typing.html#typing.ByteString>`__
+for ``typing.ByteString`` currently states:
+
+   This type represents the types ``bytes``, ``bytearray``, and
+   ``memoryview`` of byte sequences.
+
+   As a shorthand for this type, ``bytes`` can be used to annotate
+   arguments of any of the types mentioned above.
+
+Although this sentence has been in the documentation
+`since 2015 <https://github.com/python/cpython/commit/2a19d956ab92fc9084a105cc11292cb0438b322f>`__,
+the use of ``bytes`` to include these other types is not specified
+in any of the typing PEPs. Furthermore, this mechanism has a number of
+problems. It does not include all possible buffer types, and it
+makes the ``bytes`` type ambiguous in type annotations. After all,
+there are many operations that are valid on ``bytes`` objects, but
+not on ``memoryview`` objects, and it is perfectly possible for
+a function to accept ``bytes`` but not ``memoryview`` objects.
+A mypy user
+`reports <https://github.com/python/mypy/issues/12643#issuecomment-1105914159>`__
+that this shortcut has caused significant problems for the ``psycopg`` project.
+
+Kinds of buffers
+----------------
+
+The C buffer protocol supports 
+`many options <https://docs.python.org/3.10/c-api/buffer.html#buffer-request-types>`__,
+affecting strides, contiguity, and support for writing to the buffer. Some of these
+options would be useful in the type system. For example, typeshed
+currently provides separate type aliases for writable and read-only
+buffers.
+
+However, in the C buffer protocol, these options cannot be
+queried directly on the type object. The only way to figure out
+whether an object supports a writable buffer is to actually
+ask for the buffer. For some types, such as ``memoryview``,
+whether the buffer is writable depends on the instance:
+some instances are read-only and others are not. As such, we propose to
+support only whether a type implements the buffer protocol at
+all, not whether it supports more specific options such as
+writable buffers.
+
+Specification
+=============
+
+types.Buffer
+------------
+
+A new class ``types.Buffer`` will be added. It cannot be instantiated or
+subclassed, but supports the ``__instancecheck__`` and
+``__subclasscheck__`` hooks.  In CPython, these will check for the presence of the
+``bf_getbuffer`` slot in the type object:
+
+.. code-block:: pycon
```

Comment:
Freudian slip :)?
```suggestion
.. code-block:: python
```

Reply:
Umm, [no](https://pygments.org/docs/lexers/#pygments.lexers.python.PythonConsoleLexer)...

Reply:
It confuses me every time I see it too :)

Reply:
Ahh Python console. Cool! Didn't know that. Sorry!

Reply:
Yeah, you're certainly not the first person to think it was a mistake—in fact, I'm pretty sure I did myself at some point :)